client/credentials.go

```diff
@@ -0,0 +1,108 @@
+package client
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/tent/hawk-go"
+)
+
+// Credentials for taskcluster and methods to sign requests.
+type Credentials struct {
+	ClientID         string   `json:"clientId"`
+	AccessToken      string   `json:"accessToken"`
+	Certificate      string   `json:"certificate"`
+	AuthorizedScopes []string `json:"authorizedScopes"`
+}
+
+// PayloadHash creates payload hash calculator for given content-type
+func PayloadHash(contentType string) hash.Hash {
+	a := hawk.Auth{
+		Credentials: hawk.Credentials{
+			Hash: sha256.New,
+		},
+	}
+	return a.PayloadHash(contentType)
+}
+
+type certificate struct {
+	Version   int      `json:"version"`
+	Scopes    []string `json:"scopes"`
+	Start     int64    `json:"start"`
+	Expiry    int64    `json:"expiry"`
+	Seed      string   `json:"seed"`
+	Signature string   `json:"signature"`
+	Issuer    string   `json:"issuer,omitempty"`
+}
+
+type ext struct {
+	Certificate      *certificate `json:"certificate,omitempty"`
+	AuthorizedScopes *[]string    `json:"authorizedScopes,omitempty"`
+}
+
+func (c *Credentials) newAuth(method, url string, h hash.Hash) (*hawk.Auth, error) {
+	// Create a hawk auth
+	a, err := hawk.NewURLAuth(url, &hawk.Credentials{
+		ID:   c.ClientID,
+		Key:  c.AccessToken,
+		Hash: sha256.New,
+	}, 0)
+	if err != nil {
+		return nil, err
+	}
+	a.Method = method
```

Comment:
@jonasfj OK I realised my upstream fix was actually incorrect, and it needs to be fixed in this library, so actually, you'll need to refactor a bit. You need a nonce in the case of signing a request, and you shouldn't have a nonce in the case of signing a request.

So you need to split this up and call `hawk.NewURLAuth` inside `SignURL` and `hawk.NewRequestAuth` inside `SignRequest` instead of exclusively using `hawk.NewURLAuth` in both.